### DISCUSSION TITLE
Generazione automatica fattura, stripe

### DISCUSSION THREAD

**Discussion by Jean1999** [fattureincloud/api#318] [Q&A] [answered]
Buonasera,
ho da poco acquistato fatture in cloud e vorrei integrarlo con stripe in modo che quando ricevo un pagamento tramite stripe si generi automaticamente la fattura in fatture in cloud
E' possibile farlo? 
Se si, c'è una guida specifica essendo stripe uno dei principali gateway di pagamento?
Grazie in anticipo

**Comment by emajo** [comment]
Ciao @Jean1999, purtroppo non abbiamo una guida per integrare stripe con Fatture in Cloud, non dovrebbe essere però molto complicato, i passaggi che mi aspetto essere necessari sono:
1. Creazione della fattura con i dati relativi prima che il cliente inizi il processo di pagamento (quindi salvataggio dei payments come not_paid)
2. Dopo che il cliente completa il processo di pagamento verrà inviato un webhook da stripe di tipo payment succeeded
3. A questo la fattura può essere modificata settando i payments a `status:paid` 

Grazie,
Emanuele.

**Comment by Jean1999** [comment]
Grazie mille della risposta
I passaggi li avevo chiesti a chat gpt (non sono un programmatore) tuttavia
per fare quanto scritto ho necessità che mi vengano attivati i webhooks su
fatture in cloud, giusto?

Andrea

Il Gio 11 Mag 2023, 09:32 Emanuele Saccomandi ***@***.***> ha
scritto:

**Reply by valmoz** [reply]
Buongiorno, no.
Come può leggere sulla [documentazione dei Webhooks](https://developers.fattureincloud.it/docs/webhooks/), questa funzionalità serve a ricevere una notifica quando qualcosa succede su Fatture in Cloud (ad esempio, quando viene creata o modificata una fattura). Non è questo il suo caso d'uso.

Analizzando le sue necessità, direi che dovrà usare invece i [webhooks di Stripe](https://stripe.com/docs/webhooks?locale=it-IT), e gestire gli eventi pagamento per creare la fattura su Fatture in Cloud.

Prima di tutto, le suggerirei di controllare sul nostro [AppStore](https://apps.fattureincloud.it/apps) se esistono delle integrazioni che potrebbero fare al caso suo.

In caso contrario, il mio consiglio è quello di provare a cercare un programmatore freelance in grado di supportarla nella creazione di questa integrazione; se, in alternativa, desidera cimentarsi con lo sviluppo dell'integrazione in autonomia, allora proveremo a supportarla come possibile in questo sviluppo.

Prima di tutto abbiamo però la necessità di capire quali strumenti ha a disposizione per sviluppare la nuova integrazione. Sta utilizzando qualche tipo di software, come ad esempio ecommerce o programmi sviluppati da terzi?

Grazie
Mauro

**Comment by Jean1999** [comment]
Praticamente non vorrei usare software di ecommerce ma integrare
direttamente stripe con fatture in cloud

Ho già chiesto a un programmatore e mi ha detto che sicuramente si può fare
ma lui non conosce fatture in cloud. Inoltre fa lavoretti nel tempo libero
quindi magari mi ci metterebbe un mese.
Non conosco altri programmatori e inoltre si pone il problema che devo
comunque dare l'accesso a questa persona al mio account di fatture in cloud

I passaggi che ho capito devo effettuare sono la creazione del webhook su
stripe per esempio quando viene registrato un pagamento concluso con
successo e successivamente devo collegarlo a fatture in cloud tramite API.
La parte critica, se vuole posso inviarle gli screen passo passo, è quando
stripe mi chiede l'url di fatture in cloud e fatture in cloud mi chiede
l'url di stripe.

Su fatture in cloud sono andato su API e creato una nuova applicazione e ho
generato un token

Per il resto non ho capito come procedere e non ho trovato guide per
farlo...

Vi ringrazio molto se potrete aiutarmi voi

Andrea

Il Gio 11 Mag 2023, 10:53 Mauro Valota ***@***.***> ha
scritto:

**Reply by valmoz** [reply, marked as answer]
Buongiorno Andrea.
Il problema è che non esiste la possibilità di integrare Stripe e Fatture in Cloud nativamente, è necessario aggiungere un terzo componente nel mezzo, che si occupi di ricevere gli eventi da stripe e applicare le modifiche su FIC.

Guardando su AppStore, ho trovato alcune applicazioni che potrebbero aiutarla in ciò di cui ha bisogno:
- [Make](https://apps.fattureincloud.it/apps-40-make-(formerly-integromat))
- [Automyo](https://apps.fattureincloud.it/apps-54-automyo)

queste applicazioni consentono di gestire il flusso di dati tra diverse piattaforme, tra cui sicuramente Fatture in Cloud. Se esse forniscono anche l'integrazione per Stripe (non ho controllato), allora sono anche in grado di generare fatture da eventi stripe. Le consiglio di indagare questi prodotti prima di procedere.

Se resta comunque deciso a sviluppare un'integrazione in proprio, allora dovrà sviluppare un componente da zero; è un processo relativamente complesso, che quindi potrebbe richiedere una certa dimestichezza con la tecnologia e la programmazione. Per questo motivo le suggerirei prima di verificare le soluzioni qui sopra, e solo dopo considerare lo sviluppo in proprio.

Per la sua casistica, io prenderei in considerazione la possibilità di rilasciare il suo componente su un servizio Cloud, in particolare i servizi serverless. In alternativa potrebbe anche installarlo su un server, ma potrebbe richiedere una configurazione più complessa.

Ad esempio, potrebbe decidere di utilizzare un servizio cloud tra questi:
- [Heroku](https://www.heroku.com/)
- [AWS Lambda](https://aws.amazon.com/it/lambda/)
- [Google Cloud Functions](https://cloud.google.com/functions)
- [Microsoft Azure Functions](https://azure.microsoft.com/it-it/products/functions)

La lista è parziale, questi sono solo quelli che ho utilizzato io nel passato.

Utilizzando questi servizi, dovrà scegliere un linguaggio di programmazione e sviluppare un componente che deve essere in grado di:
- esporre un endpoint REST per ricevere le notifiche da Stripe
- elaborare le informazioni necessarie
- contattare le API di Fatture in Cloud per creare la fattura

Per questa parte in particolare, le nostre [SDK](https://developers.fattureincloud.it/docs/sdks/) potrebbero esserle d'aiuto.

Ad esempio, se dovesse scegliere AWS Lambda oppure Google Cloud Functions potrebbe creare una funzione Python che espone un URL (che quindi potrebbe essere usato come target dal Webhook Stripe) e che utilizza le nostre API per creare la fattura.

Grazie
Mauro

**Comment by Jean1999** [comment]
Ieri sera avevo già visto Make e provato a fare l'integrazione ma anche lì
non sono riuscito...

Riguardo la seconda casistica pensavo fosse più semplice invece ammetto di
non avere le competenze per fare questa cosa da solo.

Io sarei disposto a pagare un programmatore nel caso, lei sarebbe
disponibile ?

Il giorno gio 11 mag 2023 alle ore 12:51 Mauro Valota <
***@***.***> ha scritto:

**Reply by valmoz** [reply]
Buongiorno.
Purtroppo non mi è possibile accettare la sua richiesta. 

Il mio consiglio è quello di contattare il supporto di Make, sono convinto che ci sia un modo per utilizzare il loro tool per applicare la sua integrazione.

In alternativa posso passarle i contatti di un programmatore con cui abbiamo collaborato per la nostra applicazione Prestashop, potrebbe provare a contattarlo.
Qui il suo sito web: https://www.websuvius.it/

Grazie 
Mauro
